*** ../../setuptools-0.6c9/pkg_resources.py	2008-09-24 12:10:45.000000000 -0500
--- ../../Enstaller/pkg_resources.py	2008-11-14 11:18:34.000000000 -0600
***************
*** 24,44 ****
  from os import open as os_open
  
  
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
  def get_supported_platform():
      """Return this platform's maximum compatible version.
  
--- 24,29 ----
***************
*** 61,85 ****
      return plat
  
  
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
  __all__ = [
      # Basic resource access and distribution/entry point discovery
      'require', 'run_script', 'get_provider',  'get_distribution',
--- 46,51 ----
***************
*** 311,331 ****
  
  
  
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
  class IResourceProvider(IMetadataProvider):
      """An object that provides access to package resources"""
  
--- 277,282 ----
***************
*** 356,372 ****
  
  
  
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
  class WorkingSet(object):
      """A collection of active distributions on sys.path (or a similar list)"""
  
--- 307,312 ----
***************
*** 509,516 ****
  
          while requirements:
              req = requirements.pop(0)   # process dependencies breadth-first
!             if req in processed:
!                 # Ignore cyclic or redundant dependencies
                  continue
              dist = best.get(req.key)
              if dist is None:
--- 449,457 ----
  
          while requirements:
              req = requirements.pop(0)   # process dependencies breadth-first
!             if req in processed or req.key == 'setuptools':
!                 # Ignore cyclic or redundant dependencies or dependencies
!                 #  on setuptools which enstaller replaces.
                  continue
              dist = best.get(req.key)
              if dist is None:
***************
*** 610,618 ****
          return distributions, error_info
  
  
- 
- 
- 
      def require(self, *requirements):
          """Ensure that distributions matching `requirements` are activated
  
--- 551,556 ----
***************
*** 646,659 ****
  
  
  
- 
- 
- 
- 
- 
- 
- 
- 
  class Environment(object):
      """Searchable snapshot of distributions on a search path"""
  
--- 584,589 ----
***************
*** 817,823 ****
  
  
  
- 
  class ResourceManager:
      """Manage resource extraction and packages"""
      extraction_path = None
--- 747,752 ----
***************
*** 889,905 ****
  
  
  
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
      def get_cache_path(self, archive_name, names=()):
          """Return absolute location in cache for `archive_name` and `names`
  
--- 818,823 ----
***************
*** 925,946 ****
  
  
  
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
      def postprocess(self, tempname, filename):
          """Perform any platform-specific postprocessing of `tempname`
  
--- 843,848 ----
***************
*** 963,987 ****
  
  
  
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
      def set_extraction_path(self, path):
          """Set the base path where resources will be extracted to, if needed.
  
--- 865,870 ----
***************
*** 1099,1110 ****
      return name.replace('-','_')
  
  
- 
- 
- 
- 
- 
- 
  class NullProvider:
      """Try to implement resources and metadata for arbitrary PEP 302 loaders"""
  
--- 982,987 ----
***************
*** 1224,1233 ****
              path, base = os.path.split(path)
  
  
- 
- 
- 
- 
  class DefaultProvider(EggProvider):
      """Provides access to package resources in the filesystem"""
  
--- 1101,1106 ----
***************
*** 1267,1274 ****
  empty_provider = EmptyProvider()
  
  
- 
- 
  class ZipProvider(EggProvider):
      """Resource support for zips and eggs"""
  
--- 1140,1145 ----
***************
*** 1412,1438 ****
  
  
  
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
  class FileMetadata(EmptyProvider):
      """Metadata handler for standalone PKG-INFO files
  
--- 1283,1288 ----
***************
*** 1461,1479 ****
  
  
  
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
  class PathMetadata(DefaultProvider):
      """Metadata provider for egg directories
  
--- 1311,1316 ----
***************
*** 2005,2013 ****
  
  
  
- 
- 
- 
  class Distribution(object):
      """Wrap an actual or potential sys.path entry w/metadata"""
      def __init__(self,
--- 1842,1847 ----
***************
*** 2087,2095 ****
                  )
      version = property(version)
  
- 
- 
- 
      #@property
      def _dep_map(self):
          try:
--- 1921,1926 ----
***************
*** 2195,2217 ****
          """Return the EntryPoint object for `group`+`name`, or ``None``"""
          return self.get_entry_map(group).get(name)
  
! 
! 
! 
! 
! 
! 
! 
! 
! 
! 
! 
! 
! 
! 
! 
! 
! 
  
      def insert_on(self, path, loc = None):
          """Insert self.location in path before its nearest parent directory"""
--- 2026,2054 ----
          """Return the EntryPoint object for `group`+`name`, or ``None``"""
          return self.get_entry_map(group).get(name)
  
!     def is_non_release(self):
!         """ Checks the parsed version spec of the current package and returns
!         true if it is not a release build. Non-release builds contain any
!         qualifier that is alphabetically before 'final' or contains an 'r' for
!         a specific revision number.
!         """
!         # For each part of the version, check for any development qualifier
!         for version_part in self.parsed_version:
!             
!             # If this part is an integer or the special final qualifier, skip it
!             if version_part.isdigit() or version_part == '*final-':
!                 continue
!             
!             # If the part sorts alphabetically before 'final' or specifies that
!             # this build was made from a specific revision number, this is a
!             # non-release build. Otherwise, this is a release build
!             if version_part < '*final' or version_part == '*r':
!                 return True
!             else:
!                 return False
!             
!         #Fall-back return incase the loop exits
!         return False
  
      def insert_on(self, path, loc = None):
          """Insert self.location in path before its nearest parent directory"""
***************
*** 2255,2262 ****
  
  
      def check_version_conflict(self):
!         if self.key=='setuptools':
!             return      # ignore the inevitable setuptools self-conflicts  :(
  
          nsp = dict.fromkeys(self._get_metadata('namespace_packages.txt'))
          loc = normalize_path(self.location)
--- 2092,2099 ----
  
  
      def check_version_conflict(self):
!         if self.key=='enstaller':
!             return      # ignore the inevitable enstaller self-conflicts  :(
  
          nsp = dict.fromkeys(self._get_metadata('namespace_packages.txt'))
          loc = normalize_path(self.location)
***************
*** 2315,2341 ****
      warn(stacklevel = level+1, *args, **kw)
  
  
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
  def parse_requirements(strs):
      """Yield ``Requirement`` objects for each specification in `strs`
  
--- 2152,2157 ----
***************
*** 2404,2423 ****
  
  
  
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
- 
  class Requirement:
      def __init__(self, project_name, specs, extras):
          """DO NOT CALL THIS UNDOCUMENTED METHOD; use Requirement.parse()!"""
--- 2220,2225 ----
